File: platform/test/connection/jira/sdk/test_case.py
#!/usr/bin/env python3
"""
Connection test for Jira API via the internal jira_api SDK.

Tests proxy, verify_ssl, ca_bundle, and client cert configuration
by resolving config from AppYamlConfig and making a health-check request.
Jira Cloud uses Basic auth (email:api_token).
"""

from __future__ import annotations

import argparse
import asyncio
import base64
import json
import os
import re
import sys
import time
from pathlib import Path

_HERE = Path(__file__).resolve().parent
_PLATFORM = _HERE
while _PLATFORM.name != "platform" and _PLATFORM != _PLATFORM.parent:
    _PLATFORM = _PLATFORM.parent


def _inject_paths():
    """Add polyglot/packages_py source dirs to sys.path from platform pyproject.toml."""
    toml = _PLATFORM / "pyproject.toml"
    if not toml.exists():
        return
    for line in toml.read_text().splitlines():
        m = re.search(r'path\s*=\s*"([^"]+)"', line)
        if not m:
            continue
        pkg_path = (_PLATFORM / m.group(1)).resolve()
        src = pkg_path / "src"
        sys.path.insert(0, str(src if src.is_dir() else pkg_path))


_inject_paths()

from app_yaml_load import load_app_yaml_config
from jira_api import AsyncJiraClient


# ---------------------------------------------------------------------------
# Connection config resolution
# ---------------------------------------------------------------------------

def _resolve_connection_config(provider_cfg: dict) -> dict:
    """Resolve proxy, SSL, and cert settings from YAML + env."""
    return {
        "proxy_url": (
            provider_cfg.get("proxy_url")
            or os.environ.get("HTTPS_PROXY")
            or os.environ.get("HTTP_PROXY")
        ),
        "verify_ssl": provider_cfg.get("verify_ssl", True),
        "ca_bundle": (
            provider_cfg.get("ca_bundle")
            or os.environ.get("SSL_CERT_FILE")
            or os.environ.get("REQUESTS_CA_BUNDLE")
        ),
        "client_cert": provider_cfg.get("client_cert"),
        "client_key": provider_cfg.get("client_key"),
    }


def _resolve_credentials(provider_cfg: dict) -> tuple[str | None, str | None]:
    """Resolve Jira email and api_token from YAML + env."""
    email = (
        provider_cfg.get("email")
        or os.environ.get("JIRA_EMAIL")
    )
    api_token = (
        provider_cfg.get("api_key")
        or os.environ.get("JIRA_API_TOKEN")
    )
    return email, api_token


async def _test_connection(provider_cfg: dict, conn_cfg: dict) -> dict:
    """Test Jira API connection using the SDK."""
    email, api_token = _resolve_credentials(provider_cfg)
    if not email or not api_token:
        return {"skipped": True, "reason": "no JIRA_EMAIL/JIRA_API_TOKEN or email/api_key in config"}

    base_url = provider_cfg.get("base_url") or os.environ.get("JIRA_BASE_URL", "")
    if not base_url:
        return {"skipped": True, "reason": "no base_url or JIRA_BASE_URL configured"}

    try:
        start = time.monotonic()
        async with AsyncJiraClient(
            base_url=base_url,
            email=email,
            api_token=api_token,
        ) as client:
            # Use /rest/api/3/myself as a lightweight health check
            result = await client.get("myself")
            latency = round((time.monotonic() - start) * 1000)
            return {
                "skipped": False,
                "connected": True,
                "endpoint": "/rest/api/3/myself",
                "has_user": "accountId" in result if isinstance(result, dict) else False,
                "latency_ms": latency,
                "connection": {
                    "proxy": conn_cfg["proxy_url"] or "(none)",
                    "verify_ssl": conn_cfg["verify_ssl"],
                    "ca_bundle": conn_cfg["ca_bundle"] or "(default)",
                    "has_client_cert": bool(conn_cfg["client_cert"]),
                },
            }
    except Exception as exc:
        return {
            "skipped": False,
            "connected": False,
            "error": {"name": type(exc).__name__, "message": str(exc)},
        }


async def main() -> int:
    parser = argparse.ArgumentParser()
    parser.add_argument("--app-env", default="dev")
    parser.add_argument("--config-dir", required=True)
    args = parser.parse_args()

    app_cfg = load_app_yaml_config(config_dir=args.config_dir, app_env=args.app_env).config
    provider_cfg = app_cfg.get_nested("providers", "jira", default={}) or {}
    conn_cfg = _resolve_connection_config(provider_cfg)

    connection_test = await _test_connection(provider_cfg, conn_cfg)

    summary = {
        "service": "jira-sdk",
        "mode": "connection-test",
        "python_package": "jira_api",
        "app_yaml_initialized": True,
        "connection_config": conn_cfg,
        "connection_test": connection_test,
        "overall": "PASS" if connection_test.get("connected") else (
            "SKIP" if connection_test.get("skipped") else "FAIL"
        ),
    }
    print(json.dumps(summary, indent=2, default=str))
    return 0


if __name__ == "__main__":
    raise SystemExit(asyncio.run(main()))

File: platform/test/connection/jira/sdk/test_case.mjs
#!/usr/bin/env node

/**
 * Connection test for Jira API via the internal jira_api MJS SDK.
 *
 * Tests proxy, verify_ssl, ca_bundle, and client cert configuration
 * by resolving config from AppYamlConfig and making a health-check request.
 * Jira Cloud uses Basic auth (email:api_token).
 */

import path from 'node:path';
import { pathToFileURL } from 'node:url';

const args = process.argv.slice(2);
const readArg = (name, fallback) => {
  const idx = args.indexOf(name);
  return idx >= 0 ? args[idx + 1] : fallback;
};

const appEnv = readArg('--app-env', process.env.APP_ENV || 'dev');
const configDir = readArg('--config-dir', process.env.CONFIG_DIR);
if (!configDir) {
  console.error('Missing --config-dir');
  process.exit(2);
}

const platformRoot = path.resolve(configDir, '..', '..');
const importFile = async (absPath) => import(pathToFileURL(absPath).href);

const appYamlLoadMod = await importFile(path.join(platformRoot, 'polyglot/app_yaml_load/mjs/dist/index.js'));
const jiraApiMod = await importFile(path.join(platformRoot, 'polyglot/jira_api/mjs/src/index.mjs'));

const { config: appConfig } = await appYamlLoadMod.loadAppYamlConfig({ configDir, appEnv });
const providerConfig = appConfig.getNested(['providers', 'jira'], {}) || {};

// ---------------------------------------------------------------------------
// Connection config resolution
// ---------------------------------------------------------------------------

const connCfg = {
  proxyUrl: providerConfig.proxy_url || process.env.HTTPS_PROXY || process.env.HTTP_PROXY || null,
  verifySsl: providerConfig.verify_ssl ?? true,
  caBundlePath: providerConfig.ca_bundle || process.env.SSL_CERT_FILE || process.env.NODE_EXTRA_CA_CERTS || null,
  clientCert: providerConfig.client_cert || null,
  clientKey: providerConfig.client_key || null,
};

const email = providerConfig.email || process.env.JIRA_EMAIL || null;
const apiToken = providerConfig.api_key || process.env.JIRA_API_TOKEN || null;
const baseUrl = providerConfig.base_url || process.env.JIRA_BASE_URL || null;

// ---------------------------------------------------------------------------
// Connection test
// ---------------------------------------------------------------------------

let connectionTest;

if (!email || !apiToken) {
  connectionTest = { skipped: true, reason: 'no JIRA_EMAIL/JIRA_API_TOKEN or email/api_key in config' };
} else if (!baseUrl) {
  connectionTest = { skipped: true, reason: 'no base_url or JIRA_BASE_URL configured' };
} else {
  try {
    const start = performance.now();
    const client = new jiraApiMod.JiraFetchClient({
      baseUrl,
      email,
      apiToken,
    });

    // Use /rest/api/3/myself as a lightweight health check
    const result = await client.get('myself');
    const latency = Math.round(performance.now() - start);

    connectionTest = {
      skipped: false,
      connected: true,
      endpoint: '/rest/api/3/myself',
      hasUser: typeof result === 'object' && result !== null && 'accountId' in result,
      latency_ms: latency,
      connection: {
        proxy: connCfg.proxyUrl || '(none)',
        verify_ssl: connCfg.verifySsl,
        ca_bundle: connCfg.caBundlePath || '(default)',
        has_client_cert: Boolean(connCfg.clientCert),
      },
    };
  } catch (err) {
    connectionTest = {
      skipped: false,
      connected: false,
      error: { name: err.name, message: err.message },
    };
  }
}

const summary = {
  service: 'jira-sdk',
  mode: 'connection-test',
  mjs_package: '@mta/jira-api',
  app_yaml_initialized: true,
  connection_config: connCfg,
  connection_test: connectionTest,
  overall: connectionTest.connected ? 'PASS' : connectionTest.skipped ? 'SKIP' : 'FAIL',
};

// Redact sensitive fields before logging (CodeQL js/clear-text-logging)
const redactedSummary = { ...summary, connection_config: { ...summary.connection_config, proxyUrl: summary.connection_config.proxyUrl ? '*******' : null, clientKey: summary.connection_config.clientKey ? '*******' : null } };
console.log(JSON.stringify(redactedSummary, null, 2));
